### PR TITLE
FLW-768 Implement "Staking balance" module

### PR DESCRIPTION
**Pull request by ef1rspb**
- Balance widget
- Actions widget
Validations:
Controller account is available in the app (alert dialog to add controller account is not available once tap on any actions)
Election period is not closed (alert dialog that election period is still ongoing not available once tap on any actions)
Limit of unbonding requests (32 max)

### CODE REVIEW

File: fearless/Modules/Staking/StakingBalance/ViewModel/StakingBalanceViewModelFactory.swift [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        }
+
+        return viewModels
+    }
+
+    func createUnbondingsViewModels(
+        from balanceData: StakingBalanceData,
+        locale _: Locale
+    ) -> [UnbondingItemViewModel] {
+        let precision: Int16 = 0
```
Comment (ef1rspb) [inline fearless/Modules/Staking/StakingBalance/ViewModel/StakingBalanceViewModelFactory.swift:81]: fix in next PR

File: fearless/Modules/Staking/StakingBalance/StakingBalanceInteractor+Subscription.swift [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            executing: updateClosure,
+            failing: failureClosure,
+            options: options
+        )
+    }
+
+    func subsribeToActiveEra() {
+        let codingFactoryOperation = runtimeCodingService.fetchCoderFactoryOperation()
+
+        let activeEraWrapper: CompoundOperationWrapper<ActiveEraInfo?> =
```
Comment (ERussel) [inline fearless/Modules/Staking/StakingBalance/StakingBalanceInteractor+Subscription.swift:79]: We need to use subscription to the active era so we handle the case when unbondings/reedemables changes after a user opens the screen

File: fearless/Modules/Staking/StakingBalance/StakingBalancePresenter.swift [hunk 3]
```diff
@@ -0,0 +1,161 @@
+import SoraFoundation
+
+final class StakingBalancePresenter {
+    let interactor: StakingBalanceInteractorInputProtocol
+    let wireframe: StakingBalanceWireframeProtocol
+    let viewModelFactory: StakingBalanceViewModelFactoryProtocol
+    weak var view: StakingBalanceViewProtocol?
+    private let accountAddress: AccountAddress
+
+    var controller: AccountItem?
```
Comment (ERussel) [inline fearless/Modules/Staking/StakingBalance/StakingBalancePresenter.swift:10]: Didn't find when do we fetch this guy

File: fearless/Modules/Staking/StakingBalance/View/StakingBalanceActionsWidgetView.swift [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @objc
+    func handleActionButton(sender: UIButton) {
+        guard let actionButton = sender as? StakingBalanceActionButton else { return }
+        delegate?.didSelect(action: actionButton.action)
+    }
+}
+
+private final class StakingBalanceActionButton: UIButton {
```
Comment (ERussel) [inline fearless/Modules/Staking/StakingBalance/View/StakingBalanceActionsWidgetView.swift:54]: I think RoundedButton from SoraUI can cover this it. If we are using RoundedButton then we just set clear background and stroke colours, disable shadow and set layout type to one from mockups.

File: fearless/Modules/Staking/StakingBalance/View/StakingBalanceUnbondingItemView.swift [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    private enum Constants {
+        static let verticalInset: CGFloat = 11
+        static let iconSize: CGFloat = 32
+    }
+
+    private let transactionTypeView: UIView = {
+        UIImageView(image: R.image.iconStakingTransactionType())
+    }()
+
+    private let addressLabel: UILabel = {
```
Comment (ERussel) [inline fearless/Modules/Staking/StakingBalance/View/StakingBalanceUnbondingItemView.swift:13]: I think it should be title. There is no case where we display addresses

**Comment by antonkhvorov** [comment]
In the attached screenshot UI looks off:
- margins between rows in "Balance" widget are too low 
- Missing background with blue in the title
- margins between actions in Action widget are wrong
- separators are missing in Action widget between actions
- (maybe due to screenshot compression) separators in Balance widget and Unbonding widget are too thicc

File: fearless/Modules/Staking/StakingBalance/StakingBalanceProtocols.swift [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    func didReceive(electionStatusResult: Result<ElectionStatus?, Error>)
+    func didReceive(stashItemResult: Result<StashItem?, Error>)
+    func didReceive(fetchControllerResult: Result<(AccountItem?, AccountAddress?), Error>)
+}
+
+protocol StakingBalanceWireframeProtocol: AlertPresentable, ErrorPresentable, StakingErrorPresentable {
+    func showBondMore(from view: ControllerBackedProtocol?)
+    func showUnbond(from view: ControllerBackedProtocol?)
+    func showRedeem(from view: ControllerBackedProtocol?)
+    // TODO: add func cancel() when stashItem == nil
```
Comment (ERussel) [inline fearless/Modules/Staking/StakingBalance/StakingBalanceProtocols.swift:33]: when cancel? :)

File: fearless/Modules/Staking/StakingBalance/StakingBalancePresenter.swift [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            self.electionStatus = electionStatus
+        case .failure:
+            electionStatus = nil
+        }
+    }
+
+    func didReceive(stashItemResult: Result<StashItem?, Error>) {
+        switch stashItemResult {
+        case let .success(stashItem):
+            self.stashItem = stashItem
```
Comment (ERussel) [inline fearless/Modules/Staking/StakingBalance/StakingBalancePresenter.swift:145]: I think we can cancel here if we have nil stash item

File: fearless/Modules/Staking/StakingBalance/View/StakingBalanceUnbondingWidgetView.swift [hunk 8]
```diff
@@ -0,0 +1,77 @@
+import UIKit
+
+final class StakingBalanceUnbondingWidgetView: UIView {
+    private let backgroundView: UIView = TriangularedBlurView()
```
Comment (ERussel) [inline fearless/Modules/Staking/StakingBalance/View/StakingBalanceUnbondingWidgetView.swift:4]: Please, set blurStyle to .dark, so we are ok on iOS 12.x

File: fearless/Modules/Staking/StakingBalance/View/StakingBalanceActionsWidgetView.swift [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+        roundedBackgroundView?.backgroundColor = .clear
+        roundedBackgroundView?.strokeColor = .clear
+        roundedBackgroundView?.fillColor = .clear
+        roundedBackgroundView?.shadowColor = .clear
+
+        imageWithTitleView?.layoutType = .verticalImageFirst
+        imageWithTitleView?.iconImage = action.iconImage
+        imageWithTitleView?.titleFont = .p2Paragraph
+        imageWithTitleView?.title = action.title(for: .autoupdatingCurrent)
```
Comment (ERussel) [inline fearless/Modules/Staking/StakingBalance/View/StakingBalanceActionsWidgetView.swift:84]: I suggest to handle localization on the superview level

**Review by ERussel — APPROVED** [review]

**Pull request merged**